aleph-client/src/contract/event.rs

```diff
@@ -0,0 +1,130 @@
+//!     <Result<(), anyhow::Error>>::Ok(())
+//! };
+//! let join = tokio::spawn(listen());
+//!
+//! contract1.contract_exec0(&signed_conn, "some_method").await?;
+//! contract2.contract_exec0(&signed_conn, "some_other_method").await?;
+//!
+//! println!("Received event {:?}", rx.next().await);
+//!
+//! rx.close();
+//! join.await??;
+//!
+//! #   Ok(())
+//! # }
+//! ```
+
+use std::collections::HashMap;
+
+use anyhow::{bail, Result};
+use contract_transcode::Value;
+use futures::{channel::mpsc::UnboundedSender, StreamExt};
+
+use crate::{contract::ContractInstance, AccountId, Connection};
+
+/// Represents a single event emitted by a contract.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ContractEvent {
+    /// The address of the contract that emitted the event.
+    pub contract: AccountId,
+    /// The name of the event.
+    pub ident: Option<String>,
+    /// Data contained in the event.
+    pub data: HashMap<String, Value>,
+}
+
+/// Starts an event listening loop.
+///
+/// Will send contract event and every error encountered while fetching through the provided [UnboundedSender].
+/// Only events coming from the address of one of the `contracts` will be decoded.
+///
+/// The loop will terminate once `sender` is closed.
+pub async fn listen_contract_events(
+    conn: &Connection,
+    contracts: &[&ContractInstance],
+    sender: UnboundedSender<Result<ContractEvent>>,
+) -> Result<()> {
+    let mut block_subscription = conn.client.blocks().subscribe_finalized().await?;
+
+    while let Some(block) = block_subscription.next().await {
+        if sender.is_closed() {
+            break;
+        }
+
+        let block = block?;
+
+        for event in block.events().await?.iter() {
+            let event = event?;
+
+            if let Some(event) =
+                event.as_event::<crate::api::contracts::events::ContractEmitted>()?
+            {
+                if let Some(contract) = contracts
+                    .iter()
+                    .find(|contract| contract.address() == &event.contract)
+                {
+                    let data = zero_prefixed(&event.data);
+                    let event = contract
+                        .transcoder
+                        .decode_contract_event(&mut data.as_slice());
+
+                    sender.unbounded_send(build_event(contract.address().clone(), event))?;
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// The contract transcoder assumes there is an extra byte (that it discards) indicating the size of the data. However,
+/// data arriving through the subscription as used in this file don't have this extra byte. This function adds it.
+fn zero_prefixed(data: &[u8]) -> Vec<u8> {
+    let mut result = vec![0];
+    result.extend_from_slice(data);
+    result
+}
+
+fn build_event(address: AccountId, event_data: Result<Value>) -> Result<ContractEvent> {
```

Comment:
I would unwrap the `event_data` out of `Result` before calling this function. `build_event` shouldn't have to clean up the mess that someone else made.

Reply:
1. I don't think the cleanliness of the interface matters, because it's a private function.
2. It would add extra mangling at the call site. Cannot just move the `?` to the call site, because that's not what's supposed to happen - we're supposed to send a `Result`, so it would have to be some `if`, making the call site more complicated.

So overall, I think this is better.

Reply:
Just b/c it's a private function doesn't mean it shouldn't have a clean interface. I understand your second point though -- still, I would do that on the callsite 
```rust
sender.unbounded_send(event.map(|event_data| build_event(address, event_data)))
```
doesn't seem very complicated.

Reply:
IMO the version with `event.map` is more readable b/c it clearly "throws away" the error (signaling it's not of an interest to us in any of the contexts). So, I guess - 👎 .